element_array_ephys/export/nwb/nwb.py

```diff
@@ -0,0 +1,622 @@
+        default="units"
+    desc: str, optional
+        default="data on spiking units"
+    """
+
+    # electrode id mapping
+    mapping = get_electrodes_mapping(nwbfile.electrodes)
+
+    units_table = pynwb.misc.Units(name=name, description=desc)
+    for additional_attribute in ["cluster_quality_label", "spike_depths"]:
+        units_table.add_column(
+            name=units_query.heading.attributes[additional_attribute].name,
+            description=units_query.heading.attributes[additional_attribute].comment,
+            index=additional_attribute == "spike_depths",
+        )
+
+    clustering_query = (
+        ephys.EphysRecording * ephys.ClusteringTask & session_key & paramset_record
+    )
+
+    for unit in tqdm(
+        (clustering_query @ ephys.CuratedClustering.Unit).fetch(as_dict=True),
+        desc=f"creating units table for paramset {paramset_record['paramset_idx']}",
+    ):
+
+        probe_id, shank_num = (
+            probe.ProbeType.Electrode
+            * ephys.ProbeInsertion
+            * ephys.CuratedClustering.Unit
+            & {"unit": unit["unit"], "insertion_number": unit["insertion_number"]}
+        ).fetch1("probe", "shank")
+
+        waveform_mean = (
+            ephys.WaveformSet.PeakWaveform() & clustering_query & unit
+        ).fetch1("peak_electrode_waveform")
+
+        units_table.add_row(
+            id=unit["unit"],
+            electrodes=[mapping[(probe_id, unit["electrode"])]],
+            electrode_group=nwbfile.electrode_groups[
+                f"probe{probe_id}_shank{shank_num}"
+            ],
+            cluster_quality_label=unit["cluster_quality_label"],
+            spike_times=unit["spike_times"],
+            spike_depths=unit["spike_depths"],
+            waveform_mean=waveform_mean,
+        )
+
+    return units_table
+
+
+def add_ephys_units_to_nwb(
+    session_key: dict, nwbfile: pynwb.NWBFile, primary_clustering_paramset_idx: int = 0
+):
+    """
+    Add spiking data to NWBFile.
+
+    ephys.CuratedClustering.Unit::unit -> units.id
+    ephys.CuratedClustering.Unit::spike_times -> units["spike_times"]
+    ephys.CuratedClustering.Unit::spike_depths -> units["spike_depths"]
+    ephys.CuratedClustering.Unit::cluster_quality_label -> units["cluster_quality_label"]
+
+    ephys.WaveformSet.PeakWaveform::peak_electrode_waveform -> units["waveform_mean"]
+
+    Parameters
+    ----------
+    session_key: dict
+    nwbfile: pynwb.NWBFile
+    primary_clustering_paramset_idx: int, optional
+    """
+
+    if not ephys.ClusteringTask & session_key:
+        return
+
+    if nwbfile.electrodes is None:
+        add_electrodes_to_nwb(session_key, nwbfile)
+
+    # add additional columns to the units table
+    units_query = ephys.CuratedClustering.Unit() & session_key
+
+    for paramset_record in (
+        ephys.ClusteringParamSet & ephys.CuratedClustering & session_key
+    ).fetch("paramset_idx", "clustering_method", "paramset_desc", as_dict=True):
```

Comment:
In addition to saving units obtained for a given `paramset_idx`, we will also need to distinguish units for different rounds of curation with the `curation_id`.